src/core/structs.rs

```diff
@@ -460,6 +460,112 @@ impl<const UID_LENGTH: usize> Serializable for EncryptedTable<UID_LENGTH> {
     }
 }
 
+#[derive(Default)]
+pub struct EncryptedMultiTable<const UID_LENGTH: usize>(HashMap<Uid<UID_LENGTH>, Vec<Vec<u8>>>);
+
+impl<const UID_LENGTH: usize> EncryptedMultiTable<UID_LENGTH> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(HashMap::with_capacity(capacity))
+    }
+
+    pub fn to_encrypted_table(
+        self,
+        debug: &'static str,
+    ) -> Result<EncryptedTable<UID_LENGTH>, FindexErr> {
+        let mut table = EncryptedTable::with_capacity(self.0.len());
+
+        for (uid, mut values) in self.0 {
+            if let Some(value) = values.pop() {
+                if !values.is_empty() {
+                    return Err(FindexErr::CallBack(format!(
+                        "In {debug}, UID '{}' is associated with multiple values.",
+                        hex::encode(uid)
+                    )));
+                }
+
+                table.insert(uid, value);
+            } else {
+                return Err(FindexErr::CallBack(format!(
+                    "In {debug}, UID '{}' is associated with no values.",
+                    hex::encode(uid)
+                )));
+            }
+        }
+
+        Ok(table)
+    }
+}
+
+impl<const UID_LENGTH: usize> Deref for EncryptedMultiTable<UID_LENGTH> {
+    type Target = HashMap<Uid<UID_LENGTH>, Vec<Vec<u8>>>;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<const UID_LENGTH: usize> From<<Self as Deref>::Target> for EncryptedMultiTable<UID_LENGTH> {
+    #[inline]
+    fn from(hashmap: <Self as Deref>::Target) -> Self {
+        Self(hashmap)
+    }
+}
+
+impl<const UID_LENGTH: usize> DerefMut for EncryptedMultiTable<UID_LENGTH> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<const UID_LENGTH: usize> FromIterator<(Uid<UID_LENGTH>, Vec<Vec<u8>>)>
+    for EncryptedMultiTable<UID_LENGTH>
+{
+    fn from_iter<T: IntoIterator<Item = (Uid<UID_LENGTH>, Vec<Vec<u8>>)>>(iter: T) -> Self {
+        let hashmap = iter.into_iter().collect();
+        Self(hashmap)
+    }
+}
+
+impl<const UID_LENGTH: usize> Serializable for EncryptedMultiTable<UID_LENGTH> {
+    type Error = FindexErr;
+
+    #[inline]
+    fn length(&self) -> usize {
+        let mut length = UID_LENGTH * self.len();
+        for value in self.values() {
+            length += value.len();
+        }
+        length
+    }
```

Comment:
```suggestion
    fn length(&self) -> usize {
        to_leb128_size(self.len()) + self.iter().map(|(uid, values)| UID_LENGTH + to_leb128_size(values.len()) + values.len() * (UID_LENGTH + VALUE_LENGTH))
    }
```
where VALUE_LENGTH should be something like `EntryTableValue.length()` or `KWI_LENGTH + Keyword::HASH_LENGTH + UID_LENGTH + DemScheme::ENCRYPTION_OVERHEAD`.

Reply:
Is it wrong or just changed the style? Because this code was copied from the previous implementation without modification…

Reply:
The returned size is wrong which may lead to realocations during the serialization

Reply:
Try this:
```
    #[inline]
    fn length(&self) -> usize {
        to_leb128_len(self.len())
            + self
                .iter()
                .map(|(_uid, values)| {
                    UID_LENGTH
                        + to_leb128_len(values.len())
                        + values
                            .into_iter()
                            .map(|v| to_leb128_len(v.len()) + v.len())
                            .sum::<usize>()
                })
                .sum::<usize>()
    }
```

Reply:
I think it's ok to have re-allocation for this patch… Don't want to change too much the code… Except if you think it's really important?

Reply:
No, try that:
```
    #[inline]
    fn length(&self) -> usize {
        to_leb128_len(self.len())
            + self
                .iter()
                .map(|(_uid, values)| {
                    to_leb128_len(values.len())
                        + values
                            .into_iter()
                            .map(|v| UID_LENGTH + to_leb128_len(v.len()) + v.len())
                            .sum::<usize>()
                })
                .sum::<usize>()
    }
```